src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -2454,22 +2454,24 @@ void ShenandoahHeap::set_evacuation_reserve_quantities(bool is_valid) {
 }
 
 void ShenandoahHeap::set_concurrent_young_mark_in_progress(bool in_progress) {
-  if (has_forwarded_objects()) {
-    set_gc_state_mask(YOUNG_MARKING | UPDATEREFS, in_progress);
-  } else {
+  assert(!has_forwarded_objects(), "Young marking is not concurrent with evacuation");
+  if (!in_progress && is_concurrent_old_mark_in_progress()) {
+    assert(mode()->is_generational(), "Only generational GC has old marking");
+    // If old-marking is in progress when we turn off YOUNG_MARKING, leave MARKING (and OLD_MARKING) on
     set_gc_state_mask(YOUNG_MARKING, in_progress);
+  } else {
+    set_gc_state_mask(MARKING | YOUNG_MARKING, in_progress);
   }
-
   manage_satb_barrier(in_progress);
 }
 
 void ShenandoahHeap::set_concurrent_old_mark_in_progress(bool in_progress) {
```

Comment:
Should we assert that if `has_forwarded_objects` that the gc state mask has `UPDATEREFS` set?

Reply:
I think original code may have been wrong.  When we set_gc_state_mask(), this sets the bits specified in the argument and leaves other bits alone.  So UPDATEREFS or EVACUATION bit should have already been on.  I'll go ahead and add this assert.